File: socktest2.zig


const std = @import("std");

const c = @cImport({
  @cInclude("sys/socket.h");
  @cInclude("sys/un.h");
  @cInclude("unistd.h");
});

const SOCKET_PATH = "/tmp/socktest2.sock";


pub fn main() !void{
  std.fs.deleteFileAbsolute(SOCKET_PATH) catch {}; // delete old if exists
  
  const sock = c.socket(c.AF_UNIX, c.SOCK_DGRAM, 0); // create new TCP sock, Unix domain (TCP-like)
  if(sock == -1){
    std.debug.print("failed to create socket\n", .{});
    return error.SocketCreateFailed;
  }
  defer _ = c.close(sock);
  
  var addr: c.struct_sockaddr_un = undefined;
  addr.sun_family = c.AF_UNIX;
  
  if(SOCKET_PATH.len >= addr.sun_path.len){
    return error.SocketPathTooLong;
  }
  @memcpy(addr.sun_path[0..SOCKET_PATH.len], SOCKET_PATH);
  addr.sun_path[SOCKET_PATH.len] = 0;
  
  if(c.bind(sock, @ptrCast(&addr), @sizeOf(c.struct_sockaddr_un)) == -1){
    std.debug.print("failed to bind socket\n", .{});
    return error.SocketBindFailed;
  }
  
  std.debug.print("Zig UDP server is listening {s}\n", .{SOCKET_PATH});
  
  
  var buffer: [256]u8 = undefined;
  var sender_addr: c.struct_sockaddr_un = undefined;
  var addr_len: c.socklen_t = @sizeOf(c.struct_sockaddr_un);
  
  while(true){
    const n_bytes = c.recvfrom(
      sock,
      &buffer,
      buffer.len,
      0,
      @ptrCast(&sender_addr),
      &addr_len
    );
    
    if(n_bytes <= 0){
      std.debug.print("receive error\n", .{});
      continue;
    }
    
    const msg = buffer[0..@intCast(n_bytes)];
    std.debug.print("receive msg: {s}\n", .{msg});
    
    
    // process msg
    var reply_buffer: [128]u8 = undefined;
    var fbs = std.io.fixedBufferStream(&reply_buffer);
    var writer = fbs.writer();
    
    if(std.mem.eql(u8, msg, "hello")){
      try writer.print("Hello World (by Zig) !", .{});
    
    }else if(std.mem.startsWith(u8, msg, "factorial ")){
      const num_str = msg["factorial ".len..];
      if(std.fmt.parseInt(usize, num_str, 10)) |num|{
        
        if(num > 34){ // max n for max factorial for u128
          try writer.print("error: input too large, max is 34", .{});
        }else{
          var fact: u128 = 1;
          var i = num;
          while(i > 1) : (i -= 1){ fact *= i; }
          try writer.print("factorial({d}) = {d}", .{ num, fact });
        }
      
      }else |_|{
        try writer.print("error: invalid number", .{});
      }
    
    }else{
      try writer.print("unknown command", .{});
    }
    
    const reply = fbs.getWritten();
    
    _ = c.sendto( // send responce
      sock,
      reply.ptr,
      reply.len,
      0,
      @ptrCast(&sender_addr),
      addr_len
    );
  }
}

File: socktest2.erl
-module(socktest2).

-export([
  connect/0,
  send/1
]).


-define(SOCKET_PATH, "/tmp/socktest2.sock").
-define(SOCKET_CLIENT_PATH, "/tmp/sockclienttest2.sock").


connect() ->
  catch file:delete(?SOCKET_CLIENT_PATH), %% delete old client sock if exists
  %case socket:open(local, dgram, udp) of %% Unix domain socket UDP-like
  case socket:open(local, dgram, 0) of %% Unix domain socket UDP-like
    {ok, Socket} ->
      %%io:format("connected to Unix socket (UDP) ~s~n", [SocketPath]),
      connect(Socket);
    {error, Reason} ->
      io:format("connect err: ~p~n", [Reason]),
      error
  end.

connect(Socket) ->
  case socket:bind(Socket, #{family => local, path => ?SOCKET_CLIENT_PATH}) of
    ok -> Socket;
    {error, Reason} ->
      io:format("connect bind err: ~p~n", [Reason]),
      error
  end.


send(Message) when is_list(Message) orelse is_binary(Message) ->
  Socket = connect(),
  send(Socket, Message).

send(error, _) -> error;
send(Socket, Message) ->
  Msg = unicode:characters_to_binary(Message, utf8),
  case socket:sendto(Socket, Msg, #{family => local, path => ?SOCKET_PATH}) of
    ok ->
      
      case socket:recv(Socket, 0, 5000) of %% 5 sec timeout
        {ok, Reply} ->
          %%io:format("received: ~s~n", [Reply]),
          socket:close(Socket),
          Reply;
        
        E = {error, timeout} ->
          io:format("Receive timeout~n", []),
          socket:close(Socket),
          E;
        
        E = {error, Reason} ->
          io:format("receive err: ~p~n", [Reason]),
          socket:close(Socket),
          E
      end;
    
    E = {error, Reason} ->
      io:format("send err: ~p~n", [Reason]),
      gen_udp:close(Socket),
      E
  end.

